Logging for library authors (#36509)
dotnet/docs: docs/core/extensions/logging-library-authors.md

- The quantity value `7` is boxed into an `object`.
- The boxed quantity and description are allocated into an `object[]` to satisfy the `params` keyword.

These allocations can be avoided when logging is disabled, but that requires a guard to check if the `Information` log level is enabled:

```csharp
if (_logger.IsEnabled(LogLevel.Information))
{
    _logger.LogInformation(
        "Sold {quantity} units of {description}",
        7, "Widgets");
}
```

In situations like these, the guard prevents the allocations. All of the following extension methods have similar concerns:

- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogCritical%2A?displayProperty=nameWithType>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogDebug%2A?displayProperty=nameWithType>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogError%2A?displayProperty=nameWithType>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogInformation%2A?displayProperty=nameWithType>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogTrace%2A?displayProperty=nameWithType>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogWarning%2A?displayProperty=nameWithType>

> It's easy to forget to add the `IsEnabled` guard and it becomes repetitive. Instead, use the <xref:Microsoft.Extensions.Logging.LoggerMessageAttribute> to automatically apply the appropriate guards, calling <xref:Microsoft.Extensions.Logging.ILogger.IsEnabled%2A?displayProperty=nameWithType> with the correct <xref:Microsoft.Extensions.Logging.LogLevel>. When you use the `LoggerMessage` attribute, it ensures that the logging code is only executed if the corresponding `LogLevel` is enabled. Consider the following code:

Author: Good suggestion, I agree it's more concise.

> [!TIP]
> The source generated code can be stepped into during debugging, as it's part of the same assembly as the code that calls it.

### Expensive parameter evaluation considerations

There may be situations where evaluating parameters is expensive. Expanding upon the previous example, imagine the `description` parameter is a `string` that is expensive to compute. Perhaps the product being sold gets a friendly product description and relies on a database query, or reading from a file. In these situations, you can instruct the source generator to skip the `IsEnabled` guard and manually add the `IsEnabled` guard at the call site. This allows the user to determine where the guard is called and ensures that parameters that might be expensive to compute are only evaluated when truly needed. Consider the following code:

```csharp
using Microsoft.Extensions.Logging;

namespace Logging.LibraryAuthors;

internal static partial class LogMessages
{
    [LoggerMessage(
        Message = "Sold {quantity} of {description}",
        SkipEnabledCheck = true)]
    internal static partial void LogProductSaleDetails(
        this ILogger logger,
        int quantity,
        string description,

>         LogLevel logLevel = LogLevel.Information);

Reviewer: I'd recommend putting the level in the LoggerMessage attribute rather than as a parameter. The parameter is only for cases where the level needs to be changed dynamically which is hopefully a rare use-case.

Author: Done

---
title: Logging guidance for .NET library authors
author: IEvangelist
description: Learn how to expose logging as a library author in .NET. Follow the guidance to ensure your library is correctly exposed to consumers.
ms.author: dapine
ms.date: 08/03/2023
---

# Logging guidance for .NET library authors

As a library author, exposing logging is a great way to provide consumers with insight into the inner workings of your library. This guidance helps you expose logging in a way that is consistent with other .NET libraries and frameworks.

## Use the `ILoggerFactory` interface

> When your library exposes configuration functionality for logging, it might be tempting to accept an <xref:Microsoft.Extensions.Logging.ILogger>, but that isn't recommended as it represents a single category. An <xref:Microsoft.Extensions.Logging.ILogger%601> represents a one-to-one relationship between a category and a type. If your library allows consumers to provide one of these interfaces, it means that you're limiting the consumer to a single category, which may not be appropriate for their use case.

Reviewer: As-is this guidance may seem confusing if the library author uses DI because the guidance here is saying use ILoggerFactory and guidance elsewhere is saying use ILogger\<T> in the class constructor.

Perhaps something like this? (Feel free not to use this wording or weave in examples, I'm just trying to convey the general idea)

## When to use ILoggerFactory vs. ILogger\<T>

When writing a library that emits logs, you will need an ILogger object to record the logs. To get that object your API can either accept an ILogger\<T> parameter, or it can accept an ILoggerFactory after which you call ILoggerFactory.CreateLogger(). Which approach should be prefered?
- If you need a logging object that can be passed along to multiple classes so that all of them can emit logs, use ILoggerFactory. It is recommended that each class create logs with a separate category, named the same as the class name. In order to do this you will need the factory to create unique ILogger\<T> objects for each class that is emitting logs. Common examples of this would be public entrypoint APIs for a library or public constructors of types that internally may create helper classes.
- If you need a logging object that will only be used inside one class and never shared use ILogger\<T> where T is the type that will be producing the logs. A common example of this is a constructor for a class created by dependency injection.

If you are designing a public API that must remain stable over time, keep in mind that you might desire to refactor your internal implementation in the future. Even if a class doesn't create any internal helper types initially, that might change as the code evolves. Using ILoggerFactory will accomodate creating new ILogger\<T> objects for any new classes without changing the public API.

Instead, you should accept an <xref:Microsoft.Extensions.Logging.ILoggerFactory> and use it to create <xref:Microsoft.Extensions.Logging.ILogger%601> instances. Using an `ILoggerFactory` ensures that the logging configuration is consistent with the rest of the app and is more flexible. For more information, see [Log category](logging.md#log-category).

Logging supports filtering only by log level and category, so it's important to use the correct category for each log message. For more information, see [How filtering rules are applied](logging.md#how-filtering-rules-are-applied).

### Create a logger factory

> To create an `ILoggerFactory`, use the <xref:Microsoft.Extensions.Logging.LoggerFactory> class. This class is available from the [Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging) NuGet package. The following example shows how to create a logger factory that logs to the console:

Author: That's true, and I agree - thank you for this.

### Default with null logging

Libraries can default to _null logging_ if no `ILoggerFactory` is provided. The use of _null logging_ differs from defining types as nullable (`ILoggerFactory?`), as the types are non-null. These convenience-based types don't log anything and are essentially no-ops. Consider using any of the available abstraction types:

- <xref:Microsoft.Extensions.Logging.Abstractions.NullLogger>
- <xref:Microsoft.Extensions.Logging.Abstractions.NullLoggerFactory>
- <xref:Microsoft.Extensions.Logging.Abstractions.NullLoggerProvider>

### Dependency injection-aware libraries

If your library is dependency injection-aware, classes that require logging may accept an <xref:Microsoft.Extensions.Logging.ILogger%601> in the constructor.

:::code source="./snippets/logging/library-authors/DiExampleService.cs":::

For more information, see [.NET dependency injection](dependency-injection.md).

### Dependency injection-oblivious libraries

> If your library is dependency injection-oblivious, classes that require logging should use the configured <xref:Microsoft.Extensions.Logging.ILoggerFactory> to resolve <xref:Microsoft.Extensions.Logging.ILogger%601> instances on demand.

Reviewer: I've never heard it phrased as 'dependency injection-oblivious'. Is this a common way we say it in docs? I typically say something like "If your library isn't DI-aware" or "If your library isn't designed for DI"

Author: Yeah, I'm not sure if it's very common either - but docs are often a place where terms are coined. DI-aware and DI-oblivious are antonyms. But I agree, reading it again it's a bit clunky. I'll reword it.

## Prefer source-generated logging

Wherever your library code relies on logging APIs, for example, the `ILogger` interface, consider using source-generated logging where possible. As a general rule, when calling any of the following extension methods you should ensure that the corresponding `LogLevel` is enabled:

- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogCritical%2A>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogDebug%2A>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogError%2A>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogInformation%2A>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogTrace%2A>
- <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogWarning%2A>

Use the <xref:Microsoft.Extensions.Logging.LoggerMessageAttribute> to automatically apply the appropriate guards, calling <xref:Microsoft.Extensions.Logging.ILogger.IsEnabled%2A?displayProperty=nameWithType> with the correct <xref:Microsoft.Extensions.Logging.LogLevel>. When you use the `LoggerMessage` attribute, it ensures that the logging code is only executed if the corresponding `LogLevel` is enabled.

> Additionally, the `LoggerMessage` attribute avoids the need to use reflection at runtime, which can be a significant performance improvement.

Reviewer 2: ```suggestion
Additionally, the `LoggerMessage` attribute avoids the need to use reflection at run time, which can be a significant performance improvement.
```

---
title: Logging guidance for .NET library authors
author: IEvangelist
description: Learn how to expose logging as a library author in .NET. Follow the guidance to ensure your library is correctly exposed to consumers.
ms.author: dapine
ms.date: 08/08/2023
---

# Logging guidance for .NET library authors

As a library author, exposing logging is a great way to provide consumers with insight into the inner workings of your library. This guidance helps you expose logging in a way that is consistent with other .NET libraries and frameworks. It also helps you avoid common performance bottlenecks that may not be otherwise obvious.

## When to use the `ILoggerFactory` interface

> When writing a library that emits logs, you need an <xref:Microsoft.Extensions.Logging.ILogger> object to record the logs. To get that object your API can either accept an <xref:Microsoft.Extensions.Logging.ILogger%601> parameter, or it can accept an <xref:Microsoft.Extensions.Logging.ILoggerFactory> after which you call <xref:Microsoft.Extensions.Logging.ILoggerFactory.CreateLogger%2A?displayProperty=nameWithType>. Which approach should be preferred?

Reviewer 2: ```suggestion
When writing a library that emits logs, you need an <xref:Microsoft.Extensions.Logging.ILogger> object to record the logs. To get that object, your API can either accept an <xref:Microsoft.Extensions.Logging.ILogger%601> parameter, or it can accept an <xref:Microsoft.Extensions.Logging.ILoggerFactory> after which you call <xref:Microsoft.Extensions.Logging.ILoggerFactory.CreateLogger%2A?displayProperty=nameWithType>. Which approach should be preferred?
```

> - When you need a logging object that can be passed along to multiple classes so that all of them can emit logs, use `ILoggerFactory`. It's recommended that each class creates logs with a separate category, named the same as the class. To do this, you need the factory to create unique `ILogger<TCategoryName>` objects for each class that is emitting logs. Common examples include public entry point APIs for a library or public constructors of types that internally may create helper classes.

Reviewer 2: ```suggestion
- When you need a logging object that can be passed along to multiple classes so that all of them can emit logs, use `ILoggerFactory`. It's recommended that each class creates logs with a separate category, named the same as the class. To do this, you need the factory to create unique `ILogger<TCategoryName>` objects for each class that emits logs. Common examples include public entry point APIs for a library or public constructors of types that might create helper classes internally.
```

> - When you need a logging object that's only used inside one class and never shared, use `ILogger<TCategoryName>` where `TCategoryName` is the type that is producing the logs. A common example of this is a constructor for a class created by dependency injection.

Reviewer 2: ```suggestion
- When you need a logging object that's only used inside one class and never shared, use `ILogger<TCategoryName>`, where `TCategoryName` is the type that produces the logs. A common example of this is a constructor for a class created by dependency injection.
```

## Prefer source-generated logging

The `ILogger` API supports two approaches to using the API. You can either call methods such as <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogError%2A?displayProperty=nameWithType> and <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogInformation%2A?displayProperty=nameWithType>, or you can use the logging source generator to define strongly typed logging methods. For most situations, the source generator is recommended because it offers superior performance and stronger typing. It also isolates logging-specific concerns such as message templates, IDs, and log levels from the calling code. The non-source-generated approach is primarily useful for scenarios where you are willing to give up those advantages to make the code more concise.

:::code source="snippets/logging/library-authors/LogMessages.cs":::

The preceding code:

- Defines a `partial class` named `LogMessages`, which is `static` so that it can be used to define extension methods on the `ILogger` type.
- Decorates a `LogProductSaleDetails` extension method with the `LoggerMessage` attribute and `Message` template.

> - The `LogProductSaleDetails` extends the `ILogger` and accepts a `quantity` and `description`.

Reviewer 2: ```suggestion
- Declares `LogProductSaleDetails`, which extends the `ILogger` and accepts a `quantity` and `description`.
```

> [!TIP]

> > The source generated code can be stepped into during debugging, as it's part of the same assembly as the code that calls it.

Reviewer 2: ```suggestion
> You can step into the source-generated code during debugging, because it's part of the same assembly as the code that calls it.
```

Reviewer 2: Just to avoid double "as".
